pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -57,40 +57,44 @@ func NewWatcher(sockDir string) Watcher {
 }
 
 // AddHandler registers a callback to be invoked for a particular type of plugin
-func (w *Watcher) AddHandler(handlerType string, handlerCbkFn RegisterCallbackFn) {
+func (w *Watcher) AddHandler(pluginType string, handlerCbkFn RegisterCallbackFn) {
 	w.mutex.Lock()
 	defer w.mutex.Unlock()
-	w.handlers[handlerType] = handlerCbkFn
+	w.handlers[pluginType] = handlerCbkFn
 }
 
 // Creates the plugin directory, if it doesn't already exist.
 func (w *Watcher) createPluginDir() error {
 	glog.V(4).Infof("Ensuring Plugin directory at %s ", w.path)
 	if err := w.fs.MkdirAll(w.path, 0755); err != nil {
-		return fmt.Errorf("error (re-)creating driver directory: %s", err)
+		return fmt.Errorf("error (re-)creating root %s: %v", w.path, err)
 	}
+
 	return nil
 }
 
-// Walks through the plugin directory to discover any existing plugin sockets.
-func (w *Watcher) traversePluginDir() error {
-	files, err := w.fs.ReadDir(w.path)
-	if err != nil {
-		return fmt.Errorf("error reading the plugin directory: %v", err)
-	}
-	for _, f := range files {
-		// Currently only supports flat fs namespace under the plugin directory.
-		// TODO: adds support for hierarchical fs namespace.
-		if !f.IsDir() && filepath.Base(f.Name())[0] != '.' {
-			go func(sockName string) {
-				w.watcher.Events <- fsnotify.Event{
-					Name: sockName,
-					Op:   fsnotify.Op(uint32(1)),
-				}
-			}(path.Join(w.path, f.Name()))
+// Walks through the plugin directory discover any existing plugin sockets.
+func (w *Watcher) traversePluginDir(dir string) error {
+	return w.fs.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("error accessing path: %s error: %v", path, err)
 		}
-	}
-	return nil
+		if info.IsDir() {
+			if err := w.fsWatcher.Add(path); err != nil {
+				return fmt.Errorf("failed to watch %s, err: %v", path, err)
+			}
+			return nil
+		}
+
+		go func() {
+			w.fsWatcher.Events <- fsnotify.Event{
+				Name: path,
+				Op:   fsnotify.Op(fsnotify.Create),
```

Comment:
is `fsnotify.Create` already of type `Op`, or no?

Reply:
the `fsnotify.Create` is a constant need cast to type `fsnotify.Op`

Reply:
you are right, cast is not needed here